File: backend/app/routers/usage.py
"""GET /health, GET /usage e GET /events (SSE)."""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse

from app import __version__
from app.hub import sse_bytes
from app.netutil import panel_lan_url
from app.schemas import SSE_WIRE_EXAMPLE, HealthPayload, UsagePayload

router = APIRouter(tags=["usage"])

_SSE_HEADERS = {
    "Content-Type": "text/event-stream",
    "Connection": "keep-alive",
    "Cache-Control": "no-cache",
    "X-Accel-Buffering": "no",
}

_SSE_RESPONSE_HEADERS = {
    "Content-Type": {
        "description": "Obrigatório para o cliente tratar o corpo como SSE.",
        "schema": {"type": "string", "enum": ["text/event-stream"]},
    },
    "Connection": {
        "description": "Mantém o TCP aberto (firmware e EventSource).",
        "schema": {"type": "string", "enum": ["keep-alive"]},
    },
    "Cache-Control": {
        "description": "Impede proxy/browser de cachear o stream.",
        "schema": {"type": "string", "enum": ["no-cache"]},
    },
    "X-Accel-Buffering": {
        "description": "Pede ao nginx/Vite para não bufferizar o stream.",
        "schema": {"type": "string", "enum": ["no"]},
    },
}


@router.get(
    "/health",
    response_model=HealthPayload,
    summary="Liveness",
    operation_id="get_health",
    description=(
        "Processo no ar. Inclui os caminhos do painel, mostrador, "
        "`GET /usage`, `GET /events` e o Swagger. "
        "`panel_lan` é a URL absoluta para abrir o painel em outro aparelho da rede "
        "(a placa usa isso no QR da tela Sistema)."
    ),
)
def health(request: Request) -> HealthPayload:
    port = int(request.app.state.listen_port)
    return HealthPayload(
        version=__version__,
        panel_lan=panel_lan_url(port),
        listen={"host": request.app.state.listen_host, "port": port},
        interval_s=int(request.app.state.hub.seconds),
    )


@router.get(
    "/usage",
    response_model=UsagePayload,
    summary="Cotas na hora",
    operation_id="get_usage",
    description=(
        "Consulta Claude, GPT, Cursor, OpenRouter e DeepSeek **neste request**, "
        "grava o snapshot no hub e avisa todos os clientes de `GET /events`.\n\n"
        "HTTP **200** mesmo quando uma conta vem com `ok: false`. "
        "Use no `curl`, no botão «Atualizar consumo» e no Try it out do Swagger. "
        "A placa e o `/display` **não** fazem poll neste path — eles escutam `/events`."
    ),
    response_description="Mesmo schema do evento SSE `usage`. HTTP 200 com falha parcial.",
)
async def usage(request: Request) -> UsagePayload:
    hub = request.app.state.hub
    if request.client and request.headers.get("x-vigia-device") == "esp32":
        hub.note_device(request.client.host, request.headers.get("x-vigia-screen"))
    payload = await hub.refresh()
    return UsagePayload.model_validate(payload)


@router.get(
    "/events",
    summary="Stream SSE das cotas",
    operation_id="get_events",
    description=(
        "Server-Sent Events para o firmware (ESP32/Wokwi) e o mostrador React (`/display`).\n\n"
        "**Headers da resposta**\n\n"
        "- `Content-Type: text/event-stream`\n"
        "- `Connection: keep-alive`\n"
        "- `Cache-Control: no-cache`\n"
        "- `X-Accel-Buffering: no`\n\n"
        "**Quadro**\n\n"
        "```\n"
        "event: usage\n"
        "data: <JSON UsagePayload>\n"
        "\n"
        "```\n\n"
        "O JSON em `data:` é **idêntico** ao de `GET /usage` "
        "(schema `UsagePayload`). Comentários `: ping` a cada ~15 s mantêm NAT/ESP32 acordados. "
        "O coletor consulta as APIs a cada `USAGE_INTERVAL_S` (padrão 60 s) e empurra um `usage` "
        "a todos os inscritos. `GET /usage` força um ciclo extra.\n\n"
        "O Try it out do Swagger **segura a conexão** (stream infinito) — para um JSON único use `/usage`."
    ),
    response_class=StreamingResponse,
    responses={
        200: {
            "description": "Stream infinito. Primeiro um comentário `: connected`, depois `event: usage`.",
            "headers": _SSE_RESPONSE_HEADERS,
            "content": {
                "text/event-stream": {
                    "schema": {
                        "type": "string",
                        "description": "Quadros SSE. O campo data de `event: usage` valida contra UsagePayload.",
                    },
                    "examples": {
                        "quadro": {
                            "summary": "connected + usage + ping",
                            "value": SSE_WIRE_EXAMPLE,
                        }
                    },
                }
            },
        }
    },
)
async def events(request: Request) -> StreamingResponse:
    hub = request.app.state.hub
    if request.client and request.headers.get("x-vigia-device") == "esp32":
        hub.note_device(request.client.host, request.headers.get("x-vigia-screen"))
    return StreamingResponse(
        sse_bytes(hub),
        media_type="text/event-stream",
        headers=_SSE_HEADERS,
    )

File: backend/src/routers/usage.ts
import type { FastifyInstance } from "fastify";
import { VERSION } from "../version.js";
import { panelLanUrl } from "../netutil.js";
import { formatSse, sseBytes } from "../hub.js";

const SSE_HEADERS = {
  "Content-Type": "text/event-stream",
  "Cache-Control": "no-cache",
  "Connection": "keep-alive",
  "X-Accel-Buffering": "no",
};

export async function createUsageRoutes(app: FastifyInstance): Promise<void> {
  app.get("/health", async (request, reply) => {
    const hub = (app as unknown as { hub?: { seconds: number } }).hub;
    const listenHost = (app as unknown as { listenHost?: string }).listenHost ?? "0.0.0.0";
    const listenPort = (app as unknown as { listenPort?: number }).listenPort ?? 8787;
    const intervalS = hub?.seconds ?? 60;
    return {
      ok: true,
      version: VERSION,
      panel: "/",
      panel_lan: panelLanUrl(listenPort),
      display: "/display",
      usage: "/usage",
      events: "/events",
      docs: "/docs",
      listen: { host: listenHost, port: listenPort },
      interval_s: intervalS,
    };
  });

  app.get("/usage", async (request, reply) => {
    const hub = (app as unknown as { hub?: { refresh: (opts: unknown) => Promise<unknown>; noteDevice: (ip: string | null, screen: string | null) => void } }).hub;
    // note device via headers
    const device = (request.headers["x-vigia-device"] as string | undefined) ?? (request.headers["X-Vigia-Device"] as string | undefined);
    if (device === "esp32" && hub) {
      const ip = request.ip;
      const screen = (request.headers["x-vigia-screen"] as string | undefined) ?? (request.headers["x-vigia-screen"] as string | undefined) ?? null;
      // Fastify lowercases headers, so check case-insensitive
      const screenHeader = (request.headers["x-vigia-screen"] as string | undefined) ?? null;
      hub.noteDevice(ip, screenHeader);
    }
    if (!hub) return reply.code(503).send({ ok: false, error: "hub not ready" });
    const payload = await hub.refresh({ forceQuota: true });
    return payload;
  });

  app.get("/events", async (request, reply) => {
    const hub = (app as unknown as { hub?: InstanceType<typeof import("../hub.js").UsageHub> }).hub;
    if (!hub) return reply.code(503).send({ ok: false, error: "hub not ready" });
    const device = (request.headers["x-vigia-device"] as string | undefined);
    if (device === "esp32") {
      const ip = request.ip;
      const screen = (request.headers["x-vigia-screen"] as string | undefined) ?? null;
      hub.noteDevice(ip, screen);
    }

    // hijack raw response for SSE
    reply.hijack();
    const raw = reply.raw;
    raw.writeHead(200, SSE_HEADERS);
    // ensure flush if available
    if (typeof (raw as unknown as { flushHeaders?: () => void }).flushHeaders === "function") {
      try { (raw as unknown as { flushHeaders: () => void }).flushHeaders(); } catch {}
    }

    // stream via generator
    const gen = sseBytes(hub);
    let closed = false;
    const onClose = () => { closed = true; };
    raw.on("close", onClose);
    raw.on("error", onClose);
    try {
      for await (const chunk of gen) {
        if (closed) break;
        if (!raw.writableEnded && !raw.destroyed) {
          raw.write(typeof chunk === "string" ? chunk : Buffer.from(chunk as Uint8Array));
        } else break;
      }
    } catch {
      // ignore
    } finally {
      raw.off("close", onClose);
      try { raw.end(); } catch {}
    }
  });
}
